### PR TITLE
Replace pages-plugin with loader

### LINKED ISSUE #5598: Can not navigate in development (Router hangs after some time)
# Bug report

## Describe the bug

Either `Link` or `Router` stops working when doing client side routing. It seems that HMR may be interrupting the transition between pages. It occurs most frequently if the app is left idle or in the background for a bit of time (although I have experienced it happening whilst clicking around without it being idle)

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Clone this repo (https://github.com/malimccalla/next-routing-issue)
2.  Install dependencies `npm install`
3. Run the server `npm run dev`
4. Visit all pages by clicking the links
5. Go make a coffee (leave page idle for ~2mins)
6. Attempt to visit all pages by clicking the links
7. Certain links do not navigate to their respective page 😔 

## Expected behavior
I expect to be able to visit all the individual pages

## Actual behaviour

Page does not navigate to certain routes. Refreshing the page will fix it

## Screenshots

Both the "about" and "contact" links do not work (notice the HMR log on the first attempt of each route). After a page refresh they work as expected. If I leave the app idle for roughly 2mins again it will start over with the issue

next-issue

## System information
 -  mac Mojave 10.14
 - Chrome
 - next.js  v7.0.2

## Additional context

 With the console open you can see that the Router event `routeChangeStart` fires but `routeChangeComplete` never does.

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes #5598

Thanks to @malimccalla's clear reproduction I was able to pinpoint the issue that caused the router to hang. It's relatively complex but I'll try to explain.

Next.js has two commands for bundling: `next` (or programmatic with the `dev` flag) and `build`. When running `next build` we bundle up all pages into the `.next` directory and those files are used to serve pages with `next start`.

The development mode works differently and is optimized for running only part of your application. We call this on-demand-entries. The way it works is that when you boot up Next.js in dev mode we only compile Next.js's core files. Your pages are not compiled on bootup, instead, when you go to `http://localhost:3000/about` Next.js will inject `pages/about.js` (or `pages/about/index.js` depending on your dir structure) into webpack, wait for it to compile, and then finish rendering. This is done to make it faster for users to work on only part of the application without having to build the entire application. To give an example zeit.co has 375 pages, meaning that bootup would take a while.

On-demand-entries also has a mechanism called "disposal", meaning that if you were on `/about` and you go to `/blog`, it will dispose the page if it's not viewed for 60 seconds. If you ever looked at chrome devtools you have probably seen `on-demand-entries-ping` executing every 5 seconds, this is basically a keepalive so that disposal knows if a page is still active. In Next.js 8 / currently on canary this ping has been moved to a websocket so we no longer fill your devtools network tab with these pings.

The issue described in #5598 was caused by disposal kicking in and removing the page, but the client state not updating correctly to reflect this. So when disposal kicked in the page cache would be flushed, but an instance of the previous page would still exist in webpack's global instance cache, causing the new code that was being loaded not to execute.

This PR simplifies a lot of the HMR logic we had. Previously there were a lot of edge cases that were handled by replacing the component inside the router by sending a message from the hot-reloader server to the client side. I've refactored this to use `module.hot.accept` instead so that there is a single source of truth in replacing the component. Furthermore I've replaced the page registration and hot-self-accept plugins with a single webpack loader that will generate the page code, making it easier to understand/maintain.

It's probably good to note that we have integration tests for a lot of HMR cases / HMR error recovery and other edge cases. All tests are still green ✅.

### CODE REVIEW

File: packages/next/build/webpack/loaders/next-client-pages-loader.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import {loader} from 'webpack'
+import loaderUtils from 'loader-utils'
+
+export type ClientPagesLoaderOptions = {
+  absolutePagePath: string,
+  page: string
+}
+
+const nextServerlessLoader: loader.Loader = function () {
```
Comment (timneutkens) [inline packages/next/build/webpack/loaders/next-client-pages-loader.ts]: This loader replaces both `pages-plugin.js` and the `hot-self-accept-loader`. It's no longer a "self-accept" but an accept for the specific page, which is pretty much the same as what it used to do, except if an error is thrown it bubbles up correctly in this case.

File: packages/next/build/webpack/loaders/next-client-pages-loader.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          var updatedPage = require(${stringifiedAbsolutePagePath})
+          next.router.update(${stringifiedPage}, updatedPage.default || updatedPage)
+        })
+      }
+      return { page: page.default || page }
+    }]);
+  `
+}
+
+export default nextServerlessLoader
```
Comment (lucleray) [inline packages/next/build/webpack/loaders/next-client-pages-loader.ts]:
```suggestion
export default nextClientPagesLoader
```

File: packages/next/build/webpack/loaders/next-client-pages-loader.ts (hunk 1 above)
Comment (timneutkens) [inline packages/next/build/webpack/loaders/next-client-pages-loader.ts]:
```suggestion
const nextClientPagesLoader: loader.Loader = function () {
```

File: packages/next/build/index.ts [hunk 3]
```diff
@@ -75,12 +77,28 @@ export default async function build (dir: string, conf = null): Promise<void> {
       serverlessEntrypoints[errorPage] = `next-serverless-loader?${stringify(serverlessLoaderOptions)}!`
     }
 
-    entrypoints = serverlessEntrypoints
+    serverEntrypoints = serverlessEntrypoints
   }
+  
+  const clientEntrypoints: any = {}
+
+  Object.keys(pages).forEach(async (page) => {
```
Comment (andersonleite) [inline packages/next/build/index.ts]: Is async needed here?
Also, I saw for..in instead of Object.forEach in another parts of the code. Maybe would be better to use for..in if possible to keep consistency.

Comment (timneutkens) [reply packages/next/build/index.ts]: You're looking at outdated code 😄

**Review by andersonleite — APPROVED** [review]

**Comment by exogen** [comment]
@timneutkens What are the chances of being able to backport this fix to a 7.0.3 release instead of waiting for 8.0 to be ready? Is the code much different due to other 8.x changes?

**Comment by exogen** [comment]
tbh, I hesitate to even use the term "backport" considering that 7.x is very much the current, latest version, even if you may consider it the "previous" version due to your hard work on 8.0 every day 😄

Just wanna make sure we consider people using this in their day-to-day work who may not be able to do a major bump so easily, let alone wait for it to be released. Doesn't just apply to this fix, but anything that in theory could be applied to 7.x and not just 8.0

**Comment by timneutkens** [comment]
It's not easy / feasible to backport these changes at this point (it would require a tremendous amount of work), version 8 is going to have only slight changes, there's going to be no webpack / babel upgrade like the previous 2 major versions. Only a React upgrade, which is fully backwards compatible. After v8 I'm going to change our release cycle quite a bit, tldr being more release on the stable channel.

**Comment by exogen** [comment]
Sounds good!

**Comment by hadifarnoud** [comment]
is this fixed in v7.0.2? if not, what should I do since the v8 upgrade is a lot of work for us?

**Comment by ibraelillo** [comment]
I'm sorry, but I'm using Next 9.* and I'm still facing this issue. It starts to be annoying, and specially important to our team.

**Comment by roglol** [comment]
I am still facing the same issue. Some of the links just do not redirect to new pages

**Comment by timneutkens** [comment]
It’s really not useful to reply “same issue” create new issue that has a clear and concise reproduction that allows us to look into it. It’s impossible to investigate otherwise.

**Comment by kevinvugts** [comment]
> It’s really not useful to reply “same issue” create new issue that has a clear and concise reproduction that allows us to look into it. It’s impossible to investigate otherwise.

That doesn't make any sense at all. The same issue just still occurs on next.js version  ^9.1.3 after using @zeit/next-css version ^1.0.1 and importing css files.  No need for a full reproduction as this was already done in https://github.com/zeit/next.js/issues/5598
